File: Medium/300. Longest Increasing Subsequence/solution (4).java
// 300. Longest Increasing Subsequence

// Runtime: 0 ms, faster than 100.00% of Java online submissions for Longest Increasing Subsequence.

// Memory Usage: 37.7 MB, less than 34.00% of Java online submissions for Longest Increasing Subsequence.


class Solution {
    // Dynamic Programming with Binary Search
    public int lengthOfLIS(int[] nums) {
        int[] count = new int[nums.length];
        int len = 0;
        for (int num : nums) {
            int i = Arrays.binarySearch(count, 0, len, num);
            if (i < 0) {
                i = -(i + 1);
            }

            count[i] = num;
            if (i == len) {
                len++;
            }
        }

        return len;
    }
}

File: Medium/334. Increasing Triplet Subsequence/solution (4).cpp
// 334. Increasing Triplet Subsequence

// Runtime: 143 ms, faster than 39.72% of C++ online submissions for Increasing Triplet Subsequence.

// Memory Usage: 61.4 MB, less than 87.63% of C++ online submissions for Increasing Triplet Subsequence.


class Solution {
public:
    bool increasingTriplet(vector<int>& nums) {
        if (nums.size() < 3) {
            return false;
        }

        int firstSmall = numeric_limits<int>::max();
        int secondSmall = numeric_limits<int>::max();
        for (int i = 0; i < nums.size(); ++i) {
            int num = nums[i];
            if (num <= firstSmall) {
                firstSmall = num;
            } else if (num <= secondSmall) {
                secondSmall = num;
            } else {
                return true;
            }
        }

        return false;
    }
};